### PR TITLE
Improve GCP module

### PR DESCRIPTION
**Pull request by Selutario**
|Related issue|
|---|
| #8244 |

## Description

Hello team,

This is the PR for the development branch where all the improvements of the GCP module have been merged. This branch includes for `gcp-bucket`, `gcp-pubsub` and their unit tests. More details can be found in #8716.

### Changes in Core parser

The core parser has been updated to be able to read and process the new configuration needed for `gcp-bucket`. Here is an example configuration:

```
 <gcp-bucket>
    <run_on_start>no</run_on_start>
    <interval>1m</interval>
    <logging>debug</logging>

    <bucket type="access_logs">
        <name>framework-test-bucket</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <path>access_logs/</path>
        <remove_from_bucket> no </remove_from_bucket>
    </bucket>
</gcp-bucket>
```

The new `gcp-bucket` tag is used, keeping the old `gcp-pubsub` tag for pubsub integration. Changes in the parser for `gcp-bucket` do not affect `gcp-pubsub`.

It is possible to configure more than one bucket in the gcp-bucket configuration:

```
 <gcp-bucket>
    <run_on_start>no</run_on_start>
    <interval>1m</interval>
    <logging>debug</logging>

    <bucket type="access_logs">
        <name>framework-test-bucket</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <path>access_logs/</path>
        <remove_from_bucket>no</remove_from_bucket>
    </bucket>

    <bucket type="access_logs">
        <name>framework-test-bucket-2</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <remove_from_bucket>yes</remove_from_bucket>
    </bucket>

</gcp-bucket>
```

### Changes in unit tests

New unit tests for `gcp-bucket` have been created while existing unit tests for `gcp-pubsub` have been updated accordingly. Tests have been analysed by using Valgrind to ensure no memory leaks are present.

## How to test this

This module can be executed by Wazuh if we apply the corresponding configuration (like the previous example) in `ossec.conf`. It can also be executed manually using the following command:
```
/var/ossec/wodles/gcloud/gcloud.py -t access_logs -b framework-test-bucket -c /var/ossec/wodles/gcloud/credentials.json  -l 1
```
Finally, the pubsub integration can be also tested using the following command:
```
/var/ossec/wodles/gcloud/gcloud -t pubsub -p wazuh-dev-258815 -s test_gcloud_blog -c /var/ossec/wodles/gcloud/credentials.json -l 1
```

Regards,
Selu.

### CODE REVIEW

**Comment by noise-kngdm** [comment]
## Status update
After the changes included in [PR 9927](https://github.com/wazuh/wazuh/pull/9927), we had to rebase and refactor part of the development to solve the appeared conflicts.

### Tests performed
#### Pub/Sub module
So far we've tested that the pubsub module is working both executing it in the command line and configuring it using the `ossec.conf` file.
##### Command line
To test it we've executed the module multiple times to check if all parameters are working fine.
```
root@bc81c4381b16:/var/ossec# /var/ossec/wodles/gcloud/gcloud -p wazuh-project -s test_subscription -c /var/ossec/wodles/gcloud/credentials.json -l 2 -T pubsub   
gcloud_wodle: INFO: Received and acknowledged 100 messages
```  

```
root@bc81c4381b16:/var/ossec/wodles/gcloud# /var/ossec/wodles/gcloud/gcloud -p wazuh-project -s test_subscription -c /var/ossec/wodles/gcloud/credentials.json -l 2 -T pubsub  -t 10
gcloud_wodle: INFO: Received and acknowledged 100 messages
```  
 
```
root@bc81c4381b16:/var/ossec# /var/ossec/wodles/gcloud/gcloud -p wazuh-project -s test_subscription -c /var/ossec/wodles/gcloud/credentials.json -l 2 -T pubsub -m 150
gcloud_wodle: INFO: Received and acknowledged 150 messages
```  

```
root@bc81c4381b16:/var/ossec# /var/ossec/wodles/gcloud/gcloud -p wazuh-project -s test_subscription -c /var/ossec/wodles/gcloud/credentials.json -l 2 -T pubsub -m 250 -t 12
gcloud_wodle: INFO: Received and acknowledged 250 messages
```  

##### ossec.conf file
To check that all the affected parameters were adequately parsed, we added the following code snippet to the `ossec.conf` file.
```
<gcp-pubsub>
    <pull_on_start>yes</pull_on_start>
    <interval>1d</interval>
    <project_id>wazuh-project</project_id>
    <subscription_name>test_subscription</subscription_name>
    <num_threads>4</num_threads>
    <logging>debug</logging>
    <max_messages>500</max_messages>
    <credentials_file>/var/ossec/wodles/gcloud/credentials.json</credentials_file>
</gcp-pubsub>
```  
Then, to check that the parameters are being passed to the module, we temporarily added the subsequent code to it to write them into a file.
```
with open('/tmp/pubsub_test.txt,'w') as f:
    f.write(f'messages: {max_messages}\nn_threads: {n_threads}\nproject: {project}\nsubscription id: {subscription_id}\n')
```  
After restarting the wazuh manager, we checked the test file and got the expected output.
```
root@bc81c4381b16:/# cat /tmp/pubsub_test.txt
messages: 500
n_threads: 4
project: wazuh-project
subscription id: test_subscription
```  

#### Access logs module
To test the module, we uploaded the following file to a bucket:
```
GCloud test bucket logs
Test log 0 - 1633042692324
Test log 1 - 1633042752324
Test log 2 - 1633042812324
Test log 3 - 1633042872324
Test log 4 - 1633042932324
Test log 5 - 1633042992324
Test log 6 - 1633043052324
Test log 7 - 1633043112324
Test log 8 - 1633043172324
```
##### Command line
After executing the module in debug mode, we got the following output:
```
root@be54301d06ba:/var/ossec/wodles/gcloud# /var/ossec/wodles/gcloud/gcloud -T access_logs -b test-bucket -c credentials.json -l 1 
gcloud_wodle: INFO: Processing test_file.txt
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 0 - 1633042692324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 1 - 1633042752324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 2 - 1633042812324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 3 - 1633042872324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 4 - 1633042932324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 5 - 1633042992324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 6 - 1633043052324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 7 - 1633043112324", "source": "gcp_bucket"}}'"
gcloud_wodle: DEBUG: Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GCloud test bucket logs": "Test log 8 - 1633043172324", "source": "gcp_bucket"}}'"
gcloud_wodle: INFO: Updating previously processed files.
gcloud_wodle: INFO: Received 9 messages
```  
##### ossec.conf file
To test it, we used the following configuration:
```
 <gcp-bucket>
    <run_on_start>yes</run_on_start>
    <interval>1m</interval>
    <logging>debug</logging>
    <bucket type="access_logs">
        <name>test-bucket</name>
        <credentials_file>/var/ossec/wodles/gcloud/credentials.json</credentials_file>
        <only_logs_after>2021-Sep-01</only_logs_after>
        <remove_from_bucket>no</remove_from_bucket>
    </bucket>
 </gcp-bucket>
```  
After restarting the Wazuh manager and waiting for 1 minute we checked that the database had been created and it contained information regarding the test file:
```
root@1ceeee265657:/var/ossec/wodles/gcloud# sqlite3 gcloud.db 
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> select * from access_logs;
wazuh-dev-258815|test-bucket||test_file.txt|2021-10-04 09:59:25.336000+00:00
```  
#### Python unit tests
All tests are passing.
```
==================================== test session starts ====================================
platform linux -- Python 3.9.7, pytest-5.4.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/gonzz/git/wazuh/wodles/gcloud
plugins: asyncio-0.15.1, cov-2.12.0
collected 7 items                                                                           

tests/test_access_logs.py .                                                           [ 14%]
tests/test_bucket.py .                                                                [ 28%]
tests/test_integration.py ...                                                         [ 71%]
tests/test_subscriber.py .                                                            [ 85%]
tests/test_tools.py .                                                                 [100%]

===================================== 7 passed in 0.17s =====================================
```
#### C unit tests
All the C unit tests written for this development are passing too.
```
root@ubuntu20:/wazuh/src/unit_tests/build/wazuh_modules/gcp# valgrind --track-origins=yes --leak-check=full --keep-stacktraces=alloc-and-free --track-fds=yes --num-callers=20 --show-leak-kinds=definite,indirect --log-file=/UTValgrind.log --show-reachable=yes ./test_wmodules_gcp 
[==========] Running 23 test(s).
[ RUN      ] test_wm_gcp_pubsub_read_full_configuration
[       OK ] test_wm_gcp_pubsub_read_full_configuration
[ RUN      ] test_wm_gcp_pubsub_read_enabled_tag_invalid
[       OK ] test_wm_gcp_pubsub_read_enabled_tag_invalid
[ RUN      ] test_wm_gcp_pubsub_read_project_id_tag_invalid
[       OK ] test_wm_gcp_pubsub_read_project_id_tag_invalid
[ RUN      ] test_wm_gcp_pubsub_read_no_project_id_tag
[       OK ] test_wm_gcp_pubsub_read_no_project_id_tag
[ RUN      ] test_wm_gcp_pubsub_read_subscription_name_tag_invalid
[       OK ] test_wm_gcp_pubsub_read_subscription_name_tag_invalid
[ RUN      ] test_wm_gcp_pubsub_read_no_subscription_name_tag
[       OK ] test_wm_gcp_pubsub_read_no_subscription_name_tag
[ RUN      ] test_wm_gcp_pubsub_read_credentials_file_full_path
[       OK ] test_wm_gcp_pubsub_read_credentials_file_full_path
[ RUN      ] test_wm_gcp_pubsub_read_credentials_file_tag_empty
[       OK ] test_wm_gcp_pubsub_read_credentials_file_tag_empty
[ RUN      ] test_wm_gcp_pubsub_read_credentials_file_tag_too_long
[       OK ] test_wm_gcp_pubsub_read_credentials_file_tag_too_long
[ RUN      ] test_wm_gcp_pubsub_read_credentials_file_tag_realpath_error
[       OK ] test_wm_gcp_pubsub_read_credentials_file_tag_realpath_error
[ RUN      ] test_wm_gcp_pubsub_read_credentials_file_tag_file_not_found
[       OK ] test_wm_gcp_pubsub_read_credentials_file_tag_file_not_found
[ RUN      ] test_wm_gcp_pubsub_read_no_credentials_file_tag
[       OK ] test_wm_gcp_pubsub_read_no_credentials_file_tag
[ RUN      ] test_wm_gcp_pubsub_read_max_messages_tag_empty
[       OK ] test_wm_gcp_pubsub_read_max_messages_tag_empty
[ RUN      ] test_wm_gcp_pubsub_read_max_messages_tag_not_digit
[       OK ] test_wm_gcp_pubsub_read_max_messages_tag_not_digit
[ RUN      ] test_wm_gcp_pubsub_read_pull_on_start_tag_invalid
[       OK ] test_wm_gcp_pubsub_read_pull_on_start_tag_invalid
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_debug
[       OK ] test_wm_gcp_pubsub_read_logging_tag_debug
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_info
[       OK ] test_wm_gcp_pubsub_read_logging_tag_info
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_warning
[       OK ] test_wm_gcp_pubsub_read_logging_tag_warning
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_error
[       OK ] test_wm_gcp_pubsub_read_logging_tag_error
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_critical
[       OK ] test_wm_gcp_pubsub_read_logging_tag_critical
[ RUN      ] test_wm_gcp_pubsub_read_logging_tag_invalid
[       OK ] test_wm_gcp_pubsub_read_logging_tag_invalid
[ RUN      ] test_wm_gcp_pubsub_read_invalid_tag
[       OK ] test_wm_gcp_pubsub_read_invalid_tag
[ RUN      ] test_wm_gcp_pubsub_read_invalid_element
[       OK ] test_wm_gcp_pubsub_read_invalid_element
[==========] 23 test(s) run.
[  PASSED  ] 23 test(s).
```
```
root@ubuntu20:/# cat UTValgrind.log 
==40082== Memcheck, a memory error detector
==40082== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==40082== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==40082== Command: ./test_wmodules_gcp
==40082== Parent PID: 40045
==40082== 
==40082== 
==40082== FILE DESCRIPTORS: 4 open at exit.
==40082== Open file descriptor 3: /UTValgrind.log
==40082==    <inherited from parent>
==40082== 
==40082== Open file descriptor 2: /dev/pts/0
==40082==    <inherited from parent>
==40082== 
==40082== Open file descriptor 1: /dev/pts/0
==40082==    <inherited from parent>
==40082== 
==40082== Open file descriptor 0: /dev/pts/0
==40082==    <inherited from parent>
==40082== 
==40082== 
==40082== HEAP SUMMARY:
==40082==     in use at exit: 0 bytes in 0 blocks
==40082==   total heap usage: 3,186 allocs, 3,186 frees, 1,637,358 bytes allocated
==40082== 
==40082== All heap blocks were freed -- no leaks are possible
==40082== 
==40082== For lists of detected and suppressed errors, rerun with: -s
==40082== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```  

### Conclusions
The changes have been properly tested after the refactor and the branch is ready to be reviewed.

File: wodles/gcloud/buckets/access_logs.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        A list of JSON formatted events
+        """
+        # Clean and split each line in the file
+        lines = msg.replace('"', '').split("\n")
+
+        # Get the fieldnames from the first line
+        fieldnames = lines[0].split(",")
+        values = lines[1:]
+        tsv_file = csv.DictReader(values, fieldnames=fieldnames, delimiter=',')
+        return [dict(x, source='gcp_bucket') for x in tsv_file]
```
Comment (CarlosRS9) [inline wodles/gcloud/buckets/access_logs.py:57]:
```suggestion
        return [dict(x, source='gcp_bucket') for x in tsv_file]

```
Add an empty line at the bottom of the file

File: wodles/gcloud/buckets/bucket.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        try:
+            for event in self.load_information_from_file(blob.download_as_text()):
+                self.send_msg(self.format_msg(json.dumps(event)))
+                events += 1
+            if self.delete_file:
+                self.bucket.delete_blob(blob.name)
+        except google_exceptions.NotFound:
+            self.logger.warning(f'Unable to find "{blob.name}" in {self.bucket_name}')
+
+        return events
```
Comment (CarlosRS9) [inline wodles/gcloud/buckets/bucket.py:266]:
```suggestion
        return events

```

File: wodles/gcloud/integration.py [hunk 3]
```diff
@@ -10,132 +10,59 @@
...
+            s.close()
         except socket.error as e:
             if e.errno == 111:
-                logger.critical('Wazuh must be running')
+                self.logger.critical('Wazuh must be running')
                 raise e
             else:
-                logger.critical('Error sending event to Wazuh')
+                self.logger.critical('Error sending event to Wazuh')
                 raise e
```
Comment (CarlosRS9) [inline wodles/gcloud/integration.py:68]:
```suggestion
                raise e

```

File: wodles/gcloud/pubsub/subscriber.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    def check_permissions(self):
+        """Check if permissions are OK for executing the wodle."""
+        required_permissions = {'pubsub.subscriptions.consume'}
+        response = self.subscriber.test_iam_permissions(request={'resource': self.subscription_path,
+                                                                 'permissions': required_permissions})
+        if required_permissions.difference(response.permissions) != set():
+            error_message = 'ERROR: No permissions for executing the wodle from this subscription'
+            raise Exception(error_message)
+
+    def format_msg(self, msg: bytes) -> str:
```
Comment (CarlosRS9) [inline wodles/gcloud/pubsub/subscriber.py:93]: Use the format_msg function from `WazuhGCloudIntegration` instead

File: wodles/gcloud/pubsub/subscriber.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+        return len(response.received_messages)
+
+    def process_messages(self, max_messages: int = 100) -> int:
+        """Process the available messages in the subscription.
+        :param max_messages: Maximum number of messages to retrieve
+        :return: Number of processed messages
+        """
+        try:
+            self.wazuh_queue.connect(utils.ANALYSISD)
```
Comment (CarlosRS9) [inline wodles/gcloud/pubsub/subscriber.py:140]: Create a new function called `connect` in `WazuhGCloudIntegration` to handle the socket connection instead

File: wodles/gcloud/pubsub/subscriber.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                return processed_messages
+        except socket.error as e:
+            if e.errno == 111:
+                self.logger.critical('Wazuh must be running')
+                raise e
+            else:
+                self.logger.critical('Error sending event to Wazuh')
+                raise e
+        finally:
+            self.wazuh_queue.close()
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/gcloud/pubsub/subscriber.py:158]:
```suggestion
            self.wazuh_queue.close()

```

File: wodles/gcloud/pubsub/subscriber.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    def __init__(self, credentials_file: str, project: str, logger, subscription_id: str):
+        """Instantiate a WazuhGCloudSubscriber object.
+
+        Parameters
+        ----------
+        credentials_file : str
+            Path to credentials file
+        project : str
+            Project name
+        subscription_id : str
```
Comment (CarlosRS9) [inline wodles/gcloud/pubsub/subscriber.py:36]: The `logger` parameter description is missing

File: wodles/gcloud/pubsub/subscriber.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            self.send_msg(formatted_message)
+
+        ack_ids and self.subscriber.acknowledge(
+            request={'subscription': self.subscription_path, 'ack_ids': ack_ids}
+        )
+
+        return len(response.received_messages)
+
+    def process_messages(self, max_messages: int = 100) -> int:
+        """Process the available messages in the subscription.
```
Comment (CarlosRS9) [inline wodles/gcloud/pubsub/subscriber.py:114]: This docstring uses the old format. Restore this from the previous commits.

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**